torch/csrc/api/include/torch/nn/modules.h

```diff
@@ -31,3 +31,4 @@
 #include <torch/nn/modules/activation.h>
 #include <torch/nn/modules/normalization.h>
 #include <torch/nn/modules/transformerlayer.h>
+#include <torch/nn/modules/transformercoder.h>
```

Comment:
nit: we should probably call it `transformerencoder.h`. I know the name is a bit longer but it improves discovery (and also the word "coder" can mean a lot of things)

Reply:
good catch. or similarly to the python one, can we put everything as `transformer.h`?

Reply:
Ah, let me explain. We can not put everything into transformer.h, it will cause `cycle include` otherwise. This is due to the implementation design of the python version, since python doesn't have such problem.

Let me use encoder as an example, decoder has the same thing.
In C++ APIs, a `Module` class' constructor takes an `Options` class at the input and the `Options` class contains all the initialization parameters. TransformerEncoder module takes TransformerEncoderOptions in its constructor.  And within TransformerEncoderOptions,  it has a variable of TransformerEncoderLayer class since TransformerEncoder need TransformerEncoderLayer as input. If putting everything into one header file transformer.h. Then all the `Options` classes are within options/transformer.h, and all the `Module` classes are within modules/transformer.h, then options/transformer.h will need to include modules/transformer.h and vise versa (since TransformerEncoderOptions need TransformerEncoderLayer and TransformerEncoder need TransformerEncoderOptions). To avoid such cycle, we split the header files.

Reply:
Yeah, that's a good reason to have a separate header. Is there a reason why it is named "transformercoder.h" instead of "transformerencoder.h"?

Reply:
> Yeah, that's a good reason to have a separate header. Is there a reason why it is named "transformercoder.h" instead of "transformerencoder.h"?

Since there will be decoder code within this file as well. We have some external contributor working on decoder part